src/citrine/resources/gem_resource.py

```diff
@@ -0,0 +1,148 @@
+"""Top-level class for all object template objects and collections thereof."""
+from abc import ABC
+from typing import TypeVar, Type, List, Union, Optional, Iterator
+from uuid import UUID, uuid4
+
+from gemd.entity.base_entity import BaseEntity
+from gemd.entity.link_by_uid import LinkByUID
+
+from citrine.resources.data_concepts import _make_link_by_uid
+from citrine.resources.data_concepts import DataConcepts
+from citrine._rest.collection import Collection
+from citrine._session import Session
+
+# Proper generic parameter for typing?
+# It is not a single generic type, but rather
+# polymorphic for any GEM object
+GemResourceType = TypeVar('GemResourceType', bound='DataConcepts')
+
+class GemResourceCollection(Collection[GemResourceType], ABC):
+    """A collection of GEMD objects/templates of any kind."""
+
+    _path_template = 'projects/{project_id}/storables'
+    _dataset_agnostic_path_template = 'projects/{project_id}/storables'
+    _individual_key = None
+    _collection_key = None
+    _resource = GemResourceType
+
+    def __init__(self, project_id: UUID, dataset_id: UUID, session: Session):
+        self.project_id = project_id
+        self.dataset_id = dataset_id
+        self.session = session
+        DataConcepts._make_class_dict()
+
+    def build(self, data: dict):
+        """
+        Build an arbitary GEMD resource from a serialized dictionary.
+
+        This is an internal method, and should not be called directly by users.
+
+        Parameters
+        ----------
+        data: dict
+            A serialized data model object.
+
+        Returns
+        -------
+        GemResourceType
+            A data model object built from the dictionary.
+
+        """
+        return DataConcepts.get_type(data).build(data)
+
+    def get(self, uid: Union[UUID, str, LinkByUID, BaseEntity], *,
+                 scope: Optional[str] = None) -> GemResourceType:
+        """
+        Get a GEMD resource within the project by its id.
+
+        Parameters
+        ----------
+        uid: Union[UUID, str, LinkByUID, BaseEntity]
+            A representation of the object (Citrine id, LinkByUID, or the object itself)
+        scope: Optional[str]
+            [DEPRECATED] use a LinkByUID to specify a custom scope
```

Comment:
Noooo! Don't replicate the deprecated pattern! The only argument should be `uid`, and then you call `_make_link_by_uid(uid)`.